x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/allocation/TrainedModelAllocationService.java

```diff
@@ -0,0 +1,172 @@
+import org.elasticsearch.node.NodeClosedException;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.ConnectTransportException;
+import org.elasticsearch.xpack.core.ml.action.StartTrainedModelDeploymentAction;
+import org.elasticsearch.xpack.core.ml.action.CreateTrainedModelAllocationAction;
+import org.elasticsearch.xpack.core.ml.action.DeleteTrainedModelAllocationAction;
+import org.elasticsearch.xpack.core.ml.action.UpdateTrainedModelAllocationStateAction;
+import org.elasticsearch.xpack.core.ml.inference.allocation.TrainedModelAllocation;
+
+import java.util.Objects;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+public class TrainedModelAllocationService {
+
+    private static final Logger logger = LogManager.getLogger(TrainedModelAllocationService.class);
+
+    private final Client client;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+
+    public TrainedModelAllocationService(Client client, ClusterService clusterService, ThreadPool threadPool) {
+        this.client = new OriginSettingClient(client, ML_ORIGIN);
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.threadPool = Objects.requireNonNull(threadPool);
+    }
+
+    public void updateModelAllocationState(
+        UpdateTrainedModelAllocationStateAction.Request request,
+        ActionListener<AcknowledgedResponse> listener
+    ) {
+        ClusterState currentState = clusterService.state();
+        ClusterStateObserver observer = new ClusterStateObserver(currentState, clusterService, null, logger, threadPool.getThreadContext());
+        Predicate<ClusterState> changePredicate = MasterNodeChangePredicate.build(currentState);
+        DiscoveryNode masterNode = currentState.nodes().getMasterNode();
+        if (masterNode == null) {
+            logger.warn(
+                "[{}] no master known for allocation state update [{}]",
+                request.getModelId(),
+                request.getRoutingState().getState()
+            );
+            waitForNewMasterAndRetry(observer, UpdateTrainedModelAllocationStateAction.INSTANCE, request, listener, changePredicate);
+        }
+        client.execute(UpdateTrainedModelAllocationStateAction.INSTANCE, request, ActionListener.wrap(listener::onResponse, failure -> {
+            if (isMasterChannelException(failure)) {
+                logger.info(
+                    "[{}] master channel exception will retry on new master node for allocation state update [{}]",
+                    request.getModelId(),
+                    request.getRoutingState().getState()
+                );
+                waitForNewMasterAndRetry(observer, UpdateTrainedModelAllocationStateAction.INSTANCE, request, listener, changePredicate);
+                return;
+            }
+            listener.onFailure(failure);
+        }));
```

Comment:
for table updates, this is important. We don't want a spurious failure to cause the routing table to be stale and cause inference issues. So, if the failure has to do with some intermittent master node communication issue, we should retry.